<a href="https://colab.research.google.com/github/sergiocostaifes/PPCOMP_DM/blob/main/notebooks/06_labeling_states.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NB06 — Derivação de Estados e Transições

## 1. Contexto

A partir da série temporal com atributos derivados (NB05) e dos episódios críticos detectados (NB04), este notebook formaliza a estrutura supervisionada do problema em nível de janelas temporais de 5 minutos.

Na versão anterior do pipeline, o foco principal estava na rotulagem de estados operacionais por janela. Nesta etapa revisada, essa rotulagem é mantida, mas passa a servir também como base para a modelagem de transições entre estados consecutivos.

## 2. Objetivo

O objetivo deste notebook é:

- rotular cada janela temporal com um estado operacional;
- preservar a estrutura causal da série temporal;
- derivar o estado imediatamente seguinte de cada janela;
- formalizar as transições entre estados;
- construir uma base supervisionada específica para o problema de transição.

## 3. Papel no Pipeline Atualizado

Este notebook representa o ponto de mudança mais importante do pipeline.

A partir dele, o problema deixa de ser apenas:

- “qual é o estado desta janela?”

e passa a incorporar também a questão:

- “dada uma janela no estado BEFORE, ela evolui para DURING ou retorna para NORMAL?”

Assim, este notebook conecta:

- a detecção estatística de episódios (NB04);
- a engenharia de atributos (NB05);
- a modelagem supervisionada de transições (NB08 e NB09).

## 4. Definição dos Estados

Cada janela é rotulada em uma das seguintes categorias:

- **DURING**: janela pertencente a um episódio crítico;
- **BEFORE**: janela imediatamente anterior ao início de um episódio, dentro de um horizonte definido;
- **AFTER**: janela imediatamente posterior ao fim de um episódio, dentro de um horizonte definido;
- **NORMAL**: janela fora das categorias anteriores.

A precedência adotada para evitar conflitos é:

1. DURING  
2. BEFORE / AFTER  
3. NORMAL

## 5. Definição das Transições

Uma vez definidos os estados por janela, são derivadas as relações temporais entre janelas consecutivas:

\[
state_t \rightarrow state_{t+1}
\]

A partir dessa estrutura, este notebook constrói:

- `next_state`: estado da janela seguinte;
- `transition`: transição nominal entre estado atual e estado seguinte.

Em seguida, é derivada uma base específica para o problema central do artigo, restrita às janelas em que:

- `state == BEFORE`

e com foco nas transições:

- BEFORE → DURING
- BEFORE → NORMAL

## 6. Estratégia de Rotulagem Temporal

A base temporal de referência é o eixo `bucket_id`, em granularidade fixa de 5 minutos.

Para cada episódio \([start\_bucket, end\_bucket]\):

- DURING recebe as janelas do intervalo fechado;
- BEFORE recebe as `K_BEFORE` janelas imediatamente anteriores ao início;
- AFTER recebe as `K_AFTER` janelas imediatamente posteriores ao fim.

## 7. Saídas Esperadas

Ao final deste notebook, espera-se obter:

- base completa rotulada por estado;
- transições entre janelas consecutivas;
- base supervisionada específica para modelagem de transições;
- artefatos persistidos para uso nos notebooks seguintes.

## 8. Artefatos Gerados

Este notebook deve gerar:

- `window_5min_labeled.parquet`
- `transition_dataset.parquet`
- `06_labeling_states_summary.json`

## 9. Relação com as Próximas Etapas

Este notebook alimenta diretamente:

- NB08 — Modelagem e avaliação dos algoritmos de transição;
- NB09 — Consolidação do pipeline e geração das figuras finais.

## 10. Observações Metodológicas

- a causalidade temporal é preservada integralmente;
- não há uso de informação futura na construção das features;
- `next_state` é derivado apenas para formalizar a relação supervisionada entre janelas adjacentes;
- a base de transição é construída como um subconjunto semanticamente focalizado do problema original.

Este notebook consolida a passagem entre a descrição de estados operacionais e a modelagem preditiva de transições.

In [3]:
# ============================================================
# 06_labeling_states.ipynb
# Derivação de estados e transições (5-min)
# Episódios definidos por taxa (fail_rate)
# Pipeline PPCOMP_DM (Google Cluster Trace)
# ============================================================

# -----------------------------
# 0) BOOTSTRAP (Colab + Repo)
# -----------------------------
from pathlib import Path
import os
import sys
import subprocess
import importlib
import random
import json

import numpy as np
import pandas as pd

SEED = 42
random.seed(SEED)
np.random.seed(SEED)

if not Path("/content/drive/MyDrive").exists():
    from google.colab import drive
    drive.mount("/content/drive")
else:
    print("[Bootstrap] Google Drive já montado.")

REPO_DIR = Path("/content/drive/MyDrive/Mestrado/PPCOMP_DM")
GITHUB_REPO = "https://github.com/sergiocostaifes/PPCOMP_DM.git"

if not REPO_DIR.exists():
    REPO_DIR.parent.mkdir(parents=True, exist_ok=True)
    subprocess.run(["git", "clone", GITHUB_REPO, str(REPO_DIR)], check=True)
else:
    try:
        subprocess.run(["git", "-C", str(REPO_DIR), "pull"], check=True)
    except Exception as e:
        print("[Bootstrap] Aviso no git pull:", e)

os.chdir(str(REPO_DIR))
repo_str = str(REPO_DIR)
if repo_str not in sys.path:
    sys.path.insert(0, repo_str)

importlib.invalidate_caches()

from src.paths import FEATURES_PATH, REPORTS_PATH, ensure_dirs

ensure_dirs()

def log(msg: str) -> None:
    print(f"[06_labeling_states] {msg}")

# -----------------------------
# 1) INPUTS
# -----------------------------
FEATURES_FILE = FEATURES_PATH / "window_5min_features.parquet"
EPISODES_FILE = FEATURES_PATH / "episodes_detected.parquet"

assert FEATURES_FILE.exists(), f"Arquivo não encontrado: {FEATURES_FILE}"
assert EPISODES_FILE.exists(), f"Arquivo não encontrado: {EPISODES_FILE}"

df = pd.read_parquet(FEATURES_FILE)
episodes = pd.read_parquet(EPISODES_FILE)

required_df_cols = {"bucket_id", "fail_rate"}
required_ep_cols = {"start_bucket", "end_bucket"}

missing_df = required_df_cols - set(df.columns)
missing_ep = required_ep_cols - set(episodes.columns)

assert not missing_df, f"Colunas ausentes em features: {missing_df}"
assert not missing_ep, f"Colunas ausentes em episodes: {missing_ep}"

df = df.sort_values("bucket_id").reset_index(drop=True).copy()
episodes = episodes.sort_values(["start_bucket", "end_bucket"]).reset_index(drop=True).copy()

log(f"Features shape: {df.shape}")
log(f"Episodes shape: {episodes.shape}")

# -----------------------------
# 2) PARÂMETROS
# -----------------------------
K_BEFORE = 12   # 60 min
K_AFTER = 12    # 60 min

STATE_NORMAL = "NORMAL"
STATE_BEFORE = "BEFORE"
STATE_DURING = "DURING"
STATE_AFTER = "AFTER"

VALID_TRANSITION_TARGETS = {
    f"{STATE_BEFORE}->{STATE_DURING}": 1,
    f"{STATE_BEFORE}->{STATE_NORMAL}": 0,
}

# -----------------------------
# 3) VALIDAÇÃO DO EIXO TEMPORAL
# -----------------------------
is_sorted = bool(df["bucket_id"].is_monotonic_increasing)
has_unique_bucket = bool(df["bucket_id"].is_unique)

bucket_diff = df["bucket_id"].diff().dropna()
gap_counts = bucket_diff.value_counts().sort_index().to_dict()
n_gaps_gt1 = int((bucket_diff > 1).sum())

log(f"Eixo temporal ordenado por bucket_id: {is_sorted}")
log(f"bucket_id sem duplicidade: {has_unique_bucket}")
log(f"Número de lacunas internas (diff > 1): {n_gaps_gt1}")
log(f"Distribuição de diferenças entre buckets: {gap_counts}")

assert is_sorted, "bucket_id não está ordenado crescentemente."
assert has_unique_bucket, "Foram encontrados bucket_id duplicados."

# -----------------------------
# 4) INICIALIZAÇÃO DOS RÓTULOS
# -----------------------------
df["state"] = STATE_NORMAL

during_mask = np.zeros(len(df), dtype=bool)
before_mask = np.zeros(len(df), dtype=bool)
after_mask = np.zeros(len(df), dtype=bool)

bucket_values = df["bucket_id"].to_numpy()

# -----------------------------
# 5) MARCAÇÃO DURING
# -----------------------------
for _, ep in episodes.iterrows():
    s = int(ep["start_bucket"])
    e = int(ep["end_bucket"])
    in_range = (bucket_values >= s) & (bucket_values <= e)
    during_mask |= in_range

df.loc[during_mask, "state"] = STATE_DURING
log(f"DURING total: {int(during_mask.sum())}")

# -----------------------------
# 6) MARCAÇÃO BEFORE / AFTER
# -----------------------------
for _, ep in episodes.iterrows():
    s = int(ep["start_bucket"])
    e = int(ep["end_bucket"])

    before_range = (bucket_values >= s - K_BEFORE) & (bucket_values < s)
    after_range = (bucket_values > e) & (bucket_values <= e + K_AFTER)

    before_mask |= (before_range & ~during_mask)
    after_mask |= (after_range & ~during_mask)

df.loc[(df["state"] == STATE_NORMAL) & before_mask, "state"] = STATE_BEFORE
df.loc[(df["state"] == STATE_NORMAL) & after_mask, "state"] = STATE_AFTER

# -----------------------------
# 7) VALIDAÇÕES DA ROTULAGEM
# -----------------------------
dist_states = df["state"].value_counts().to_dict()

conf_before_during = int((before_mask & during_mask).sum())
conf_after_during = int((after_mask & during_mask).sum())

episodes_without_during = 0
for _, ep in episodes.iterrows():
    s = int(ep["start_bucket"])
    e = int(ep["end_bucket"])
    hits = int(((df["bucket_id"] >= s) &
                (df["bucket_id"] <= e) &
                (df["state"] == STATE_DURING)).sum())
    if hits == 0:
        episodes_without_during += 1

log(f"Distribuição de estados: {dist_states}")
log(f"Conflitos BEFORE∩DURING: {conf_before_during}")
log(f"Conflitos AFTER∩DURING: {conf_after_during}")
log(f"Episódios sem DURING: {episodes_without_during}")

assert conf_before_during == 0, "Há conflito entre BEFORE e DURING."
assert conf_after_during == 0, "Há conflito entre AFTER e DURING."
assert episodes_without_during == 0, "Há episódios sem qualquer janela DURING."

# Coerência com janelas críticas detectadas no NB04
if "duration_windows" in episodes.columns:
    expected_during = int(episodes["duration_windows"].sum())
    actual_during = int((df["state"] == STATE_DURING).sum())
    log(f"DURING esperado (soma duration_windows): {expected_during}")
    log(f"DURING observado: {actual_during}")
    assert actual_during == expected_during, (
        f"Inconsistência no total de janelas DURING: "
        f"esperado={expected_during}, observado={actual_during}"
    )

# -----------------------------
# 8) DERIVAÇÃO DE next_state E transition (1 passo)
# -----------------------------
df["next_bucket_id"] = df["bucket_id"].shift(-1)
df["next_state"] = df["state"].shift(-1)

df["is_last_window"] = df["next_bucket_id"].isna()
df["has_valid_next"] = (
    (~df["is_last_window"]) &
    ((df["next_bucket_id"] - df["bucket_id"]) == 1)
)

df["transition"] = np.where(
    df["has_valid_next"],
    df["state"] + "->" + df["next_state"],
    pd.NA
)

# -----------------------------
# 9) DATASET DE TRANSIÇÃO (VERSÃO CORRIGIDA)
# Problema binário:
#   1 = entra em DURING dentro do horizonte H
#   0 = permanece NORMAL durante todo o horizonte H
# -----------------------------
HORIZON_WINDOWS = K_BEFORE  # 12 janelas = 60 min

state_by_bucket = dict(zip(df["bucket_id"], df["state"]))

future_valid_flags = []
future_has_during_flags = []
future_all_normal_flags = []
future_path_states = []
future_first_during_offset = []

for b in df["bucket_id"].tolist():
    future_states = []
    valid_horizon = True

    for h in range(1, HORIZON_WINDOWS + 1):
        next_b = b + h
        s = state_by_bucket.get(next_b, None)
        if s is None:
            valid_horizon = False
            break
        future_states.append(s)

    if not valid_horizon:
        future_valid_flags.append(False)
        future_has_during_flags.append(False)
        future_all_normal_flags.append(False)
        future_path_states.append(pd.NA)
        future_first_during_offset.append(pd.NA)
        continue

    has_during = any(s == STATE_DURING for s in future_states)
    all_normal = all(s == STATE_NORMAL for s in future_states)

    first_during = pd.NA
    if has_during:
        for idx, s in enumerate(future_states, start=1):
            if s == STATE_DURING:
                first_during = idx
                break

    future_valid_flags.append(True)
    future_has_during_flags.append(has_during)
    future_all_normal_flags.append(all_normal)
    future_path_states.append(" | ".join(future_states))
    future_first_during_offset.append(first_during)

df["future_horizon_valid"] = future_valid_flags
df["future_has_during"] = future_has_during_flags
df["future_all_normal"] = future_all_normal_flags
df["future_path_states"] = future_path_states
df["future_first_during_offset"] = future_first_during_offset

# Elegibilidade:
# - não usar DURING nem AFTER como estado atual
# - exigir horizonte futuro completo e contíguo
# - manter apenas casos claramente positivos ou claramente negativos
eligible_mask = (
    df["state"].isin([STATE_NORMAL, STATE_BEFORE]) &
    df["future_horizon_valid"] &
    (df["future_has_during"] | df["future_all_normal"])
)

transition_df = df.loc[eligible_mask].copy()

transition_df["transition_target"] = np.where(
    transition_df["future_has_during"], 1, 0
).astype("int64")

transition_df["transition_label"] = np.where(
    transition_df["transition_target"] == 1,
    f"TO_{STATE_DURING}_WITHIN_{HORIZON_WINDOWS}_WINDOWS",
    f"STAYS_{STATE_NORMAL}_{HORIZON_WINDOWS}_WINDOWS"
)

# Campo auxiliar interpretativo
transition_df["prediction_horizon_minutes"] = HORIZON_WINDOWS * 5

# Sanidade
assert transition_df["transition_target"].isin([0, 1]).all(), \
    "Foram encontrados rótulos inválidos em transition_target."

dist_transition_target = (
    transition_df["transition_target"]
    .value_counts()
    .sort_index()
    .to_dict()
)

dist_transition_label = (
    transition_df["transition_label"]
    .value_counts()
    .to_dict()
)

dist_current_state_in_transition = (
    transition_df["state"]
    .value_counts()
    .to_dict()
)

log(f"Transition dataset shape: {transition_df.shape}")
log(f"Distribuição de transition_target: {dist_transition_target}")
log(f"Distribuição de transition_label: {dist_transition_label}")
log(f"Distribuição do estado atual no transition dataset: {dist_current_state_in_transition}")

# -----------------------------
# 10) PERSISTÊNCIA
# -----------------------------
OUT_LABELED_FILE = FEATURES_PATH / "window_5min_labeled.parquet"
OUT_TRANSITION_FILE = FEATURES_PATH / "transition_dataset.parquet"

df.to_parquet(OUT_LABELED_FILE, compression="snappy", index=False)
transition_df.to_parquet(OUT_TRANSITION_FILE, compression="snappy", index=False)

log(f"Salvo labeled: {OUT_LABELED_FILE}")
log(f"Salvo transition dataset: {OUT_TRANSITION_FILE}")

# -----------------------------
# 11) SUMMARY JSON
# -----------------------------
summary = {
    "input_features_file": str(FEATURES_FILE),
    "input_episodes_file": str(EPISODES_FILE),
    "output_labeled_file": str(OUT_LABELED_FILE),
    "output_transition_file": str(OUT_TRANSITION_FILE),

    "rows_labeled": int(len(df)),
    "cols_labeled": int(df.shape[1]),
    "rows_transition": int(len(transition_df)),
    "cols_transition": int(transition_df.shape[1]),

    "k_before": int(K_BEFORE),
    "k_after": int(K_AFTER),
    "horizon_windows": int(HORIZON_WINDOWS),
    "horizon_minutes": int(HORIZON_WINDOWS * 5),

    "is_bucket_id_sorted": bool(is_sorted),
    "is_bucket_id_unique": bool(has_unique_bucket),
    "n_internal_gaps_gt1": int(n_gaps_gt1),
    "bucket_diff_distribution": {str(k): int(v) for k, v in gap_counts.items()},

    "states_distribution": {k: int(v) for k, v in dist_states.items()},
    "transition_target_distribution": {str(k): int(v) for k, v in dist_transition_target.items()},
    "transition_label_distribution": {k: int(v) for k, v in dist_transition_label.items()},
    "transition_current_state_distribution": {
        k: int(v) for k, v in dist_current_state_in_transition.items()
    },

    "episodes_total": int(len(episodes)),
    "episodes_without_during": int(episodes_without_during),
    "conflicts_before_during": int(conf_before_during),
    "conflicts_after_during": int(conf_after_during),
}

if "duration_windows" in episodes.columns:
    summary["during_expected_from_episodes"] = int(episodes["duration_windows"].sum())
    summary["during_observed"] = int((df["state"] == STATE_DURING).sum())

summary_file = REPORTS_PATH / "06_labeling_states_summary.json"
summary_file.write_text(
    json.dumps(summary, indent=2, ensure_ascii=False),
    encoding="utf-8"
)

log(f"Resumo salvo: {summary_file}")
log("Notebook 06 finalizado com sucesso.")

# -----------------------------
# 12) AMOSTRAS DE CONFERÊNCIA
# -----------------------------
print("\n[Prévia da base rotulada]")
display(df.head())

print("\n[Prévia da base de transição]")
display(transition_df.head())

print("\n[Distribuição final de estados]")
display(pd.Series(dist_states, name="count").to_frame())

print("\n[Distribuição final de transition_target]")
display(pd.Series(dist_transition_target, name="count").to_frame())

print("\n[Distribuição final de transition_label]")
display(pd.Series(dist_transition_label, name="count").to_frame())

print("\n[Distribuição do estado atual no transition dataset]")
display(pd.Series(dist_current_state_in_transition, name="count").to_frame())

[Bootstrap] Google Drive já montado.
[06_labeling_states] Features shape: (8914, 27)
[06_labeling_states] Episodes shape: (285, 13)
[06_labeling_states] Eixo temporal ordenado por bucket_id: True
[06_labeling_states] bucket_id sem duplicidade: True
[06_labeling_states] Número de lacunas internas (diff > 1): 1
[06_labeling_states] Distribuição de diferenças entre buckets: {1.0: 8912, 2.0: 1}
[06_labeling_states] DURING total: 394
[06_labeling_states] Distribuição de estados: {'NORMAL': 5533, 'BEFORE': 1996, 'AFTER': 991, 'DURING': 394}
[06_labeling_states] Conflitos BEFORE∩DURING: 0
[06_labeling_states] Conflitos AFTER∩DURING: 0
[06_labeling_states] Episódios sem DURING: 0
[06_labeling_states] DURING esperado (soma duration_windows): 394
[06_labeling_states] DURING observado: 394
[06_labeling_states] Transition dataset shape: (6808, 41)
[06_labeling_states] Distribuição de transition_target: {0: 4812, 1: 1996}
[06_labeling_states] Distribuição de transition_label: {'STAYS_NORMAL_12_WIND

,bucket_id,bucket_start_us,n_events,n_failed,n_machines,n_collections,mean_priority,mean_req_cpus,mean_req_mem,req_cpus_presence_rate,...,next_bucket_id,next_state,is_last_window,has_valid_next,transition,future_horizon_valid,future_has_during,future_all_normal,future_path_states,future_first_during_offset
0,15,4500000000,44,10,41,29,249.522727,0.008850,0.003220,1.0,...,16.0,NORMAL,False,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>
1,16,4800000000,22,5,22,19,219.045455,0.011749,0.003593,1.0,...,17.0,NORMAL,False,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>
2,17,5100000000,28,5,27,22,279.392857,0.010309,0.004613,1.0,...,18.0,NORMAL,False,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>
3,18,5400000000,29,7,29,22,259.517241,0.007249,0.009697,1.0,...,19.0,NORMAL,False,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>
4,19,5700000000,36,9,34,27,237.027778,0.010176,0.009284,1.0,...,20.0,NORMAL,False,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>



[Prévia da base de transição]


,bucket_id,bucket_start_us,n_events,n_failed,n_machines,n_collections,mean_priority,mean_req_cpus,mean_req_mem,req_cpus_presence_rate,...,has_valid_next,transition,future_horizon_valid,future_has_during,future_all_normal,future_path_states,future_first_during_offset,transition_target,transition_label,prediction_horizon_minutes
0,15,4500000000,44,10,41,29,249.522727,0.008850,0.003220,1.0,...,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>,0,STAYS_NORMAL_12_WINDOWS,60
1,16,4800000000,22,5,22,19,219.045455,0.011749,0.003593,1.0,...,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>,0,STAYS_NORMAL_12_WINDOWS,60
2,17,5100000000,28,5,27,22,279.392857,0.010309,0.004613,1.0,...,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>,0,STAYS_NORMAL_12_WINDOWS,60
3,18,5400000000,29,7,29,22,259.517241,0.007249,0.009697,1.0,...,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>,0,STAYS_NORMAL_12_WINDOWS,60
4,19,5700000000,36,9,34,27,237.027778,0.010176,0.009284,1.0,...,True,NORMAL->NORMAL,True,False,True,NORMAL | NORMAL | NORMAL | NORMAL | NORMAL | N...,<NA>,0,STAYS_NORMAL_12_WINDOWS,60



[Distribuição final de estados]


,count
NORMAL,5533
BEFORE,1996
AFTER,991
DURING,394



[Distribuição final de transition_target]


,count
0,4812
1,1996



[Distribuição final de transition_label]


,count
STAYS_NORMAL_12_WINDOWS,4812
TO_DURING_WITHIN_12_WINDOWS,1996



[Distribuição do estado atual no transition dataset]


,count
NORMAL,4812
BEFORE,1996


## 11. Conclusão da Etapa

A rotulagem temporal das janelas foi concluída com sucesso, resultando em uma base supervisionada estruturada em quatro estados operacionais:

- NORMAL
- BEFORE
- DURING
- AFTER

A estratégia de precedência adotada garantiu consistência semântica entre os estados, evitando sobreposição entre janelas críticas e suas regiões adjacentes.

---

## 12. Importância para a Modelagem de Transições

Este notebook representa a principal transição metodológica do pipeline, ao evoluir de uma abordagem baseada em classificação de estados para uma modelagem orientada a transições temporais.

A partir da derivação de `next_state` e da construção de trajetórias futuras, torna-se possível formular o problema supervisionado de forma mais expressiva:

- antecipação da entrada em estado crítico;
- distinção entre comportamento estável e degradação iminente.

Essa abordagem permite capturar não apenas o estado atual do sistema, mas sua dinâmica de evolução no tempo.

---

## 13. Encaminhamento do Pipeline

Com a base de transição construída, o pipeline segue para:

➡️ NB08 — Modelagem e avaliação de algoritmos  
➡️ NB09 — Consolidação e geração de resultados finais  

---

## 14. Achados Experimentais (Execução Atual)

### Entrada e consistência temporal

- Base de features: **8914 janelas**, **27 colunas**
- Base de episódios: **285 registros**
- `bucket_id` ordenado e sem duplicidade
- **1 lacuna interna** identificada (diff = 2)

Distribuição de diferenças:
- diff = 1: 8912 ocorrências
- diff = 2: 1 ocorrência

---

### Rotulagem dos estados

Distribuição final:

- NORMAL: **5533**
- BEFORE: **1996**
- AFTER: **991**
- DURING: **394**

Validações:

- Conflitos BEFORE ∩ DURING: **0**
- Conflitos AFTER ∩ DURING: **0**
- Episódios sem DURING: **0**

Coerência com episódios:

- DURING esperado: **394**
- DURING observado: **394**

---

### Dataset de transição

A base de transição foi construída considerando um horizonte de **12 janelas (60 minutos)**.

Dimensão:

- Total de registros: **6808**
- Total de atributos: **41 colunas**

Distribuição do alvo:

- Classe 0 (permanece normal): **4812**
- Classe 1 (entra em DURING): **1996**

Distribuição percentual aproximada:

- Classe 0: ~70,7%
- Classe 1: ~29,3%

Estados de origem:

- NORMAL: 4812
- BEFORE: 1996

---

### Interpretação metodológica

O problema supervisionado passa a ser formulado como:

> “Dada a janela atual, o sistema entrará em estado crítico dentro de 60 minutos?”

Essa formulação:

- preserva causalidade temporal;
- evita vazamento de informação futura;
- captura dinâmicas de antecipação;
- amplia o poder explicativo do modelo.

---

## 15. Conclusão Experimental

O Notebook 06 consolidou com sucesso a transição entre:

- detecção estatística de episódios (NB04);
- engenharia de atributos (NB05);
- modelagem supervisionada de transições.

A base rotulada está consistente, validada e alinhada com a estrutura de episódios.

A base de transição apresenta:

- volume adequado de dados;
- distribuição balanceada entre classes;
- coerência temporal e semântica;

estando pronta para uso na etapa de modelagem e avaliação de algoritmos.

Este notebook estabelece o núcleo metodológico do trabalho.